File: lua/test/recursive_observation_contract_test.lua
-- FUTURE-PARITY-BACKLOG.14.4.6 — shared Lua recursive-observation admission.
--
-- Ordinary Lua discovery and canonical CI execute this exact Lua-5.1-compatible
-- consumer once with PUC Lua and once with LuaJIT. The implementation remains
-- private: no facade, schema, semantic/MCP, CLI, or README surface is added.

local json = require("linkedspec.json")
local linkedspec = require("linkedspec")
local recognition_runtime = require("linkedspec.recognition_transaction_runtime")

local assertions = 0
local failures = {}

local function check(condition, label)
  assertions = assertions + 1
  if not condition then failures[#failures + 1] = label end
end

local function check_equal(actual, expected, label)
  check(actual == expected, label .. ": expected " .. tostring(expected) .. ", got " .. tostring(actual))
end

local function check_same_json(actual, expected, label)
  check_equal(json.encode(actual), json.encode(expected), label)
end

local function capture(operation)
  local ok, value = pcall(operation)
  return ok, value
end

local function expect_error(needle, operation, label)
  local ok, captured = capture(operation)
  check_equal(ok, false, label .. " rejects")
  if not ok then
    check(
      tostring(captured):find(needle, 1, true) ~= nil,
      label .. " preserves " .. needle .. ": got " .. tostring(captured)
    )
  end
end

local function capture_bindings(operation)
  local original = recognition_runtime.bind_observation
  local bindings = {}
  recognition_runtime.bind_observation = function(ctx, rule_label, target, scope)
    original(ctx, rule_label, target, scope)
    bindings[#bindings + 1] = json.decode(json.encode(ctx.harrays[target]))
  end
  local ok, captured = capture(operation)
  recognition_runtime.bind_observation = original
  return ok, captured, bindings
end

local function compile_source(source)
  local parsed = linkedspec.parse_spec_with_staged_user_function_definitions(source)
  linkedspec.validate_spec(parsed)
  return parsed, linkedspec.compile_spec(parsed)
end

local function all_objects(value, result)
  result = result or {}
  if type(value) ~= "table" then return result end
  result[#result + 1] = value
  for _, child in pairs(value) do
    if type(child) == "table" then all_objects(child, result) end
  end
  return result
end

local function objects_with_kind(value, kind)
  local result = {}
  for _, object in ipairs(all_objects(value)) do
    if object.kind == kind then result[#result + 1] = object end
  end
  return result
end

local function expected_observation(options)
  local selected_match = json.null
  if options.selected_start ~= nil then
    selected_match = json.harray({
      source_id = "input",
      start = options.selected_start,
      ["end"] = options.selected_end,
      provenance = "match",
    })
  end
  local accepted_exit = json.null
  if options.accepted_exit ~= nil then
    accepted_exit = json.harray({ source_id = "input", offset = options.accepted_exit })
  end
  return json.harray({
    source_id = "input",
    rule_label = options.rule_label,
    invocation_id = options.invocation_id,
    parent_invocation_id = options.parent_invocation_id or json.null,
    entry_position = json.harray({ source_id = "input", offset = options.entry_offset }),
    selected_match = selected_match,
    accepted_exit = accepted_exit,
    outcome = options.outcome,
    diagnostic = options.diagnostic or json.null,
  })
end

local authored_source = [[Top::
 I {
  value = observe_recognition(observation, call(Child))
  return(array(value, observation))
 }

Child::AND
 /🙂/
 E { return(false) }
]]

-- Dedicated lowering and static/effect policy.
do
  local _, compiled = compile_source(authored_source)
  local top = linkedspec.compiled_spec_to_json(compiled).rules_by_label.Top
  local observations = objects_with_kind(top, "observe_recognition")
  check_equal(#observations, 1, "one dedicated observation node")
  if #observations == 1 then
    check_equal(observations[1].target, "observation", "observation target")
    check_equal(observations[1].rule, "Child", "observation static child")
  else
    check(false, "observation target unavailable")
    check(false, "observation static child unavailable")
  end
  local eager_calls = 0
  for _, object in ipairs(all_objects(top)) do
    if object.kind == "call" and object.name == "Child" then eager_calls = eager_calls + 1 end
  end
  check_equal(eager_calls, 0, "observation child is not eagerly called")
  local serialized = json.encode(linkedspec.compiled_spec_to_json(compiled))
  local _, serialized_count = serialized:gsub('"kind":"observe_recognition"', "")
  check_equal(serialized_count, 1, "serialized observation node is exact once")

  local fixtures = {
    {
      [[Top::
 I { value = observe_recognition(observation["nested"], call(Child)) }
Child::AND
 /x/
]],
      "source_location_recursive_observation_target",
      "nested observation target",
    },
    {
      [[Top::
 I { value = observe_recognition(observation, dynamic_child) }
Child::AND
 /x/
]],
      "source_location_recursive_observation_operand",
      "dynamic observation operand",
    },
    {
      [[Top:: I { value = observe_recognition(observation, call(Missing)) }]],
      "source_location_recursive_observation_operand",
      "missing observation rule",
    },
    {
      [[Top::
 I {
  tx = recognition_checkpoint()
  matched = recognize_once(tx, call(Observer))
  recognition_rollback(tx)
  return(matched)
 }
Observer: I { value = observe_recognition(observation, call(Child)); return(value) }
Child::AND /x/
]],
      "recognition_effect_forbidden:binding_write",
      "rule observation effect",
    },
    {
      [[fn inspect() { value = observe_recognition(observation, call(Child)); return(value) }
Top::
 I {
  tx = recognition_checkpoint()
  matched = recognize_once(tx, call(Observer))
  recognition_rollback(tx)
  return(matched)
 }
Observer: I { return(inspect()) }
Child::AND /x/
]],
      "recognition_effect_forbidden:binding_write",
      "function observation effect",
    },
  }
  for _, fixture in ipairs(fixtures) do
    expect_error(fixture[2], function() compile_source(fixture[1]) end, fixture[3])
  end
end

-- Native and reconstructed carriers preserve false payloads and detached records.
do
  local parsed, compiled = compile_source(authored_source)
  local expected = json.array({
    false,
    expected_observation({
      rule_label = "Child",
      invocation_id = 2,
      parent_invocation_id = 1,
      entry_offset = 0,
      selected_start = 0,
      selected_end = 1,
      accepted_exit = 1,
      outcome = "accepted",
    }),
  })
  local ok, first = capture(function()
    return linkedspec.runtime_parse(linkedspec.runtime_engine(compiled), "🙂").value
  end)
  check_equal(ok, true, "native observation executes")
  if ok then
    check_same_json(first, expected, "native false payload and observation")
    first[2].entry_position.offset = 99
    first[2].selected_match.start = 99
    check_same_json(
      linkedspec.runtime_parse(linkedspec.runtime_engine(compiled), "🙂").value,
      expected,
      "native observation is detached"
    )
  end

  local reconstructed = linkedspec.spec_ast.from_json(
    "SpecFile",
    json.decode(json.encode(linkedspec.spec_ast.to_json(parsed)))
  )
  linkedspec.validate_spec(reconstructed)
  local reconstructed_ok, reconstructed_value = capture(function()
    return linkedspec.runtime_parse(
      linkedspec.runtime_engine(linkedspec.compile_spec(reconstructed)),
      "🙂"
    ).value
  end)
  check_equal(reconstructed_ok, true, "reconstructed observation executes")
  if reconstructed_ok then
    check_same_json(reconstructed_value, expected, "reconstructed observation")
  end
end

-- Failure, zero-regex, and action-edge cursor semantics.
do
  local _, failed = compile_source([[Top::
 I { value = observe_recognition(observation, call(Missing)); return(array(value, observation)) }
Missing::AND
 /z/
]])
  local failed_ok, failed_value = capture(function()
    return linkedspec.runtime_parse(linkedspec.runtime_engine(failed), "x").value
  end)
  check_equal(failed_ok, true, "failed observation executes")
  if failed_ok then
    check_same_json(failed_value, json.array({
      json.null,
      expected_observation({
        rule_label = "Missing", invocation_id = 2, parent_invocation_id = 1,
        entry_offset = 0, outcome = "failed",
      }),
    }), "failed observation")
  end

  local _, zero = compile_source([[Top::
 I { value = observe_recognition(observation, call(Coordinator)); return(array(value, observation)) }
Coordinator:
 I { return("coordinated") }
]])
  local zero_ok, zero_value = capture(function()
    return linkedspec.runtime_parse(linkedspec.runtime_engine(zero), "").value
  end)
  check_equal(zero_ok, true, "zero-regex observation executes")
  if zero_ok then
    check_same_json(zero_value, json.array({
      "coordinated",
      expected_observation({
        rule_label = "Coordinator", invocation_id = 2, parent_invocation_id = 1,
        entry_offset = 0, accepted_exit = 0, outcome = "accepted",
      }),
    }), "zero-regex observation")
  end

  local _, edge = compile_source([[Top::
 /a/ -> Top {
  value = observe_recognition(observation, call(Child))
  return(array(value, observation))
 }
Child::AND
 /b/
 /c/
 -> Child[0] { first = match_text() }
 -> Child[1] { return(array(entry_text(), entry_start_pos())) }
]])
  local edge_ok, edge_value = capture(function()
    return linkedspec.runtime_parse(linkedspec.runtime_engine(edge), "abc").value
  end)
  check_equal(edge_ok, true, "action-edge observation executes")
  if edge_ok then
    check_same_json(edge_value, json.array({
      json.array({ "a", 0 }),
      expected_observation({
        rule_label = "Child", invocation_id = 2, parent_invocation_id = 1,
        entry_offset = 1, selected_start = 2, selected_end = 3,
        accepted_exit = 3, outcome = "accepted",
      }),
    }), "action-edge observation")
  end
end

-- Ordinary nested cutoff stays ordinary; explicit direct/mutual observation is typed.
do
  local _, ordinary = compile_source([[Top:: I { value = observe_recognition(observation, call(Child)); return(array(value, observation)) }
Child: I { nested = call(Child); return("guarded") }
]])
  local ordinary_ok, ordinary_value = capture(function()
    return linkedspec.runtime_parse(linkedspec.runtime_engine(ordinary), "").value
  end)
  check_equal(ordinary_ok, true, "ordinary nested cutoff executes")
  if ordinary_ok then
    check_same_json(ordinary_value, json.array({
      "guarded",
      expected_observation({
        rule_label = "Child", invocation_id = 2, parent_invocation_id = 1,
        entry_offset = 0, accepted_exit = 0, outcome = "accepted",
      }),
    }), "ordinary nested cutoff is not observed rejection")
  end

  local _, direct = compile_source([[Top:: I { value = observe_recognition(top_observation, call(DirectRecur)); return(value) }
DirectRecur: I { value = observe_recognition(observation, call(DirectRecur)); return(value) }
]])
  local direct_ok, direct_error, direct_bindings = capture_bindings(function()
    return linkedspec.runtime_parse(linkedspec.runtime_engine(direct), "x")
  end)
  check_equal(direct_ok, false, "direct observed recursion rejects")
  check(
    tostring(direct_error):find("source_location_nonprogress_direct_recursion", 1, true) ~= nil,
    "direct observed recursion preserves typed error: got " .. tostring(direct_error)
  )
  check_same_json(direct_bindings[1] or json.null, expected_observation({
    rule_label = "DirectRecur", invocation_id = 3, parent_invocation_id = 2,
    entry_offset = 0, outcome = "rejected",
    diagnostic = "source_location_nonprogress_direct_recursion",
  }), "direct rejection binds before propagation")

  local _, mutual = compile_source([[Top:: I { value = observe_recognition(top_observation, call(MutualA)); return(value) }
MutualA: I { value = observe_recognition(observation_a, call(MutualB)); return(value) }
MutualB: I { value = observe_recognition(observation_b, call(MutualA)); return(value) }
]])
  local mutual_ok, mutual_error, mutual_bindings = capture_bindings(function()
    return linkedspec.runtime_parse(linkedspec.runtime_engine(mutual), "x")
  end)
  check_equal(mutual_ok, false, "mutual observed recursion rejects")
  check(
    tostring(mutual_error):find("source_location_nonprogress_mutual_recursion", 1, true) ~= nil,
    "mutual observed recursion preserves typed error: got " .. tostring(mutual_error)
  )
  check_same_json(mutual_bindings[1] or json.null, expected_observation({
    rule_label = "MutualA", invocation_id = 4, parent_invocation_id = 3,
    entry_offset = 0, outcome = "rejected",
    diagnostic = "source_location_nonprogress_mutual_recursion",
  }), "mutual rejection binds before propagation")
end

-- Aborted observation propagates the original typed runtime diagnostic.
do
  local _, aborted = compile_source([[Top::
 I { value = observe_recognition(observation, call(AbortChild)); return(value) }
AbortChild:
 I { recognition_commit(missing) }
]])
  local aborted_ok, aborted_error, aborted_bindings = capture_bindings(function()
    return linkedspec.runtime_parse(linkedspec.runtime_engine(aborted), "")
  end)
  check_equal(aborted_ok, false, "aborted observation rejects")
  check(
    tostring(aborted_error):find("recognition_token_expected", 1, true) ~= nil,
    "aborted observation preserves original typed error: got " .. tostring(aborted_error)
  )
  check_same_json(aborted_bindings[1] or json.null, expected_observation({
    rule_label = "AbortChild", invocation_id = 2, parent_invocation_id = 1,
    entry_offset = 0, outcome = "aborted",
  }), "aborted observation binds before propagation")
end

-- Generated-plan carrier.
do
  local _, compiled = compile_source(authored_source)
  local generated_ok, generated_value = capture(function()
    return linkedspec.execute_generated_parser_v2(
      compiled,
      linkedspec.build_generated_rule_plan(compiled),
      "🙂",
      "recursive-observation/lua.spec"
    )
  end)
  check_equal(generated_ok, true, "generated-plan observation executes")
  if generated_ok then
    check_same_json(generated_value, json.array({
      false,
      expected_observation({
        rule_label = "Child", invocation_id = 2, parent_invocation_id = 1,
        entry_offset = 0, selected_start = 0, selected_end = 1,
        accepted_exit = 1, outcome = "accepted",
      }),
    }), "generated-plan observation")
  end
end

-- Independently loaded emitted carrier.
do
  local _, compiled = compile_source(authored_source)
  local identity = "recursive-observation/lua-emitted.spec"
  local emitted = linkedspec.emit_lua_source_v2(compiled, identity)
  local loader = loadstring or load
  local chunk, failure = loader(emitted, "@recursive_observation_generated.lua")
  check(chunk ~= nil, "emitted observation source loads")
  if chunk == nil then
    failures[#failures + 1] = tostring(failure)
  else
    local generated_module = chunk()
    local emitted_ok, emitted_value = capture(function() return generated_module.execute("🙂") end)
    check_equal(emitted_ok, true, "emitted observation executes")
    if emitted_ok then
      check_same_json(emitted_value, json.array({
        false,
        expected_observation({
          rule_label = "Child", invocation_id = 2, parent_invocation_id = 1,
          entry_offset = 0, selected_start = 0, selected_end = 1,
          accepted_exit = 1, outcome = "accepted",
        }),
      }), "emitted observation")
    end
    check_equal(generated_module.metadata().source_identity, identity, "emitted observation identity")
  end
end

if #failures == 0 then
  io.stdout:write(
    "Lua recursive-observation contract: ", assertions,
    " assertions passed on ", linkedspec.runtime_implementation(), "\n"
  )
else
  io.stderr:write(
    "Lua recursive-observation contract: ", #failures,
    " of ", assertions, " assertions failed on ",
    linkedspec.runtime_implementation(), "\n"
  )
  for _, message in ipairs(failures) do io.stderr:write("- ", message, "\n") end
  os.exit(1)
end

File: julia/test/recursive_observation_contract_test.jl
# FUTURE-PARITY-BACKLOG.14.4.5 — Julia recursive-observation admission.
#
# Ordinary Julia discovery and canonical CI execute this exact consumer through
# repository-local project data. The implementation remains private: this adds
# no facade export, descriptor/schema revision, semantic/MCP projection, or CLI
# surface.

module JuliaRecursiveObservationAdmission

using JSON3
using LinkedSpecJulia
using Test

const JULIA_RECURSIVE_OBSERVATION_SOURCE = """Top::
 I {
  value = observe_recognition(observation, call(Child))
  return(array(value, observation))
 }

Child::AND
 /🙂/
 E { return(false) }
"""

function _julia_recursive_observation_compile(source::AbstractString)
    parsed = parse_spec_with_staged_user_function_definitions(source)
    validate_spec(parsed)
    return parsed, compile_spec(parsed)
end

function _julia_recursive_observation_objects(value)
    objects = Dict{String,Any}[]
    if value isa AbstractDict
        object = Dict{String,Any}(String(key) => child for (key, child) in pairs(value))
        push!(objects, object)
        for child in values(object)
            append!(objects, _julia_recursive_observation_objects(child))
        end
    elseif value isa AbstractVector
        for child in value
            append!(objects, _julia_recursive_observation_objects(child))
        end
    end
    return objects
end

function _julia_recursive_observation_kind(value, kind::AbstractString)
    return [
        object for object in _julia_recursive_observation_objects(value)
        if get(object, "kind", nothing) == kind
    ]
end

function _julia_recursive_observation_expected(;
    rule_label::AbstractString,
    invocation_id::Int,
    parent_invocation_id::Union{Nothing,Int},
    entry_offset::Int,
    selected_match::Union{Nothing,Tuple{Int,Int}} = nothing,
    accepted_exit::Union{Nothing,Int} = nothing,
    outcome::AbstractString,
    diagnostic::Union{Nothing,AbstractString} = nothing,
)
    return Dict{String,Any}(
        "source_id" => "input",
        "rule_label" => String(rule_label),
        "invocation_id" => invocation_id,
        "parent_invocation_id" => parent_invocation_id,
        "entry_position" => Dict{String,Any}(
            "source_id" => "input",
            "offset" => entry_offset,
        ),
        "selected_match" => selected_match === nothing ? nothing : Dict{String,Any}(
            "source_id" => "input",
            "start" => selected_match[1],
            "end" => selected_match[2],
            "provenance" => "match",
        ),
        "accepted_exit" => accepted_exit === nothing ? nothing : Dict{String,Any}(
            "source_id" => "input",
            "offset" => accepted_exit,
        ),
        "outcome" => String(outcome),
        "diagnostic" => diagnostic === nothing ? nothing : String(diagnostic),
    )
end

function _julia_recursive_observation_expect_error(
    needle::AbstractString,
    operation::Function,
)
    captured = try
        operation()
        nothing
    catch error
        error
    end
    @test captured !== nothing
    captured === nothing && return
    @test occursin(String(needle), sprint(showerror, captured))
end

function _julia_recursive_observation_execute_emitted(
    compiled::CompiledSpec,
    identity::AbstractString,
    input::AbstractString,
)
    emitted = emit_julia_source_v2(compiled, identity)
    host = Module(gensym(:JuliaRecursiveObservationHost))
    Base.include_string(host, emitted, "recursive_observation_generated.jl")
    parser = Base.invokelatest(() -> getfield(host, :LinkedSpecGeneratedParser))
    execute = Base.invokelatest(() -> getfield(parser, :execute))
    metadata = Base.invokelatest(() -> getfield(parser, :metadata))
    value = Base.invokelatest(execute, String(input))
    source_identity = Base.invokelatest(metadata).source_identity
    return value, source_identity
end

@testset "Julia recursive-observation contract" begin
    @testset "dedicated non-eager node and static policy are exact" begin
        _, compiled = _julia_recursive_observation_compile(
            JULIA_RECURSIVE_OBSERVATION_SOURCE,
        )
        top = compiled_rule(compiled, "Top")
        action = to_json(only(top.lifecycle_action_payloads).action_ast)
        observation = only(_julia_recursive_observation_kind(
            action,
            "observe_recognition",
        ))
        @test observation["target"] == "observation"
        @test observation["rule"] == "Child"
        @test isempty([
            object for object in _julia_recursive_observation_objects(action)
            if get(object, "kind", nothing) == "call" &&
               get(object, "name", nothing) == "Child"
        ])
        serialized = JSON3.write(to_json(compiled))
        @test length(collect(eachmatch(r"\"kind\":\"observe_recognition\"", serialized))) == 1

        fixtures = [
            (
                """Top:: I { value = observe_recognition(observation[\"nested\"], call(Child)) } Child::AND /x/""",
                "source_location_recursive_observation_target",
            ),
            (
                """Top:: I { value = observe_recognition(observation, dynamic_child) } Child::AND /x/""",
                "source_location_recursive_observation_operand",
            ),
            (
                """Top:: I { value = observe_recognition(observation, call(Missing)) }""",
                "source_location_recursive_observation_operand",
            ),
            (
                """Top::
 I {
  tx = recognition_checkpoint()
  matched = recognize_once(tx, call(Observer))
  recognition_rollback(tx)
  return(matched)
 }
Observer:
 I { value = observe_recognition(observation, call(Child)); return(value) }
Child::AND
 /x/
""",
                "recognition_effect_forbidden:binding_write",
            ),
            (
                """fn inspect() {
 value = observe_recognition(observation, call(Child))
 return(value)
}
Top::
 I {
  tx = recognition_checkpoint()
  matched = recognize_once(tx, call(Observer))
  recognition_rollback(tx)
  return(matched)
 }
Observer: I { return(inspect()) }
Child::AND
 /x/
""",
                "recognition_effect_forbidden:binding_write",
            ),
        ]
        for (source, diagnostic) in fixtures
            _julia_recursive_observation_expect_error(
                diagnostic,
                () -> _julia_recursive_observation_compile(source),
            )
        end
    end

    @testset "native and reconstructed carriers preserve false and detachment" begin
        parsed, compiled = _julia_recursive_observation_compile(
            JULIA_RECURSIVE_OBSERVATION_SOURCE,
        )
        expected = Any[
            false,
            _julia_recursive_observation_expected(
                rule_label = "Child",
                invocation_id = 2,
                parent_invocation_id = 1,
                entry_offset = 0,
                selected_match = (0, 1),
                accepted_exit = 1,
                outcome = "accepted",
            ),
        ]
        engine = LinkedSpecRuntimeEngine(compiled)
        first = runtime_parse(engine, "🙂").value
        @test first == expected
        first[2]["entry_position"]["offset"] = 99
        first[2]["selected_match"]["start"] = 99
        @test runtime_parse(engine, "🙂").value == expected

        normalized = JSON3.read(JSON3.write(to_json(parsed)), Dict{String,Any})
        reconstructed = from_json(SpecFile, normalized)
        validate_spec(reconstructed)
        @test runtime_parse(
            LinkedSpecRuntimeEngine(compile_spec(reconstructed)),
            "🙂",
        ).value == expected
    end

    @testset "failed zero-regex and action-edge cursor semantics are exact" begin
        _, failed = _julia_recursive_observation_compile("""Top::
 I { value = observe_recognition(observation, call(Missing)); return(array(value, observation)) }
Missing::AND
 /z/
""")
        @test runtime_parse(LinkedSpecRuntimeEngine(failed), "x").value == Any[
            nothing,
            _julia_recursive_observation_expected(
                rule_label = "Missing",
                invocation_id = 2,
                parent_invocation_id = 1,
                entry_offset = 0,
                outcome = "failed",
            ),
        ]

        _, zero = _julia_recursive_observation_compile("""Top::
 I { value = observe_recognition(observation, call(Coordinator)); return(array(value, observation)) }
Coordinator:
 I { return(\"coordinated\") }
""")
        @test runtime_parse(LinkedSpecRuntimeEngine(zero), "").value == Any[
            "coordinated",
            _julia_recursive_observation_expected(
                rule_label = "Coordinator",
                invocation_id = 2,
                parent_invocation_id = 1,
                entry_offset = 0,
                accepted_exit = 0,
                outcome = "accepted",
            ),
        ]

        _, edge = _julia_recursive_observation_compile("""Top::
 /a/ -> Top {
  value = observe_recognition(observation, call(Child))
  return(array(value, observation))
 }
Child::AND
 /b/
 /c/
 -> Child[0] { first = match_text() }
 -> Child[1] { return(array(entry_text(), entry_start_pos())) }
""")
        @test runtime_parse(LinkedSpecRuntimeEngine(edge), "abc").value == Any[
            Any["a", 0],
            _julia_recursive_observation_expected(
                rule_label = "Child",
                invocation_id = 2,
                parent_invocation_id = 1,
                entry_offset = 1,
                selected_match = (2, 3),
                accepted_exit = 3,
                outcome = "accepted",
            ),
        ]
    end

    @testset "direct and mutual nonprogress keep typed diagnostics" begin
        _, ordinary = _julia_recursive_observation_compile("""Top:: I { value = observe_recognition(observation, call(Child)); return(array(value, observation)) }
Child: I { nested = call(Child); return(\"guarded\") }
""")
        @test runtime_parse(LinkedSpecRuntimeEngine(ordinary), "").value == Any[
            "guarded",
            _julia_recursive_observation_expected(
                rule_label = "Child",
                invocation_id = 2,
                parent_invocation_id = 1,
                entry_offset = 0,
                accepted_exit = 0,
                outcome = "accepted",
            ),
        ]

        _, direct = _julia_recursive_observation_compile("""Top:: I { value = observe_recognition(top_observation, call(DirectRecur)); return(value) }
DirectRecur: I { value = observe_recognition(observation, call(DirectRecur)); return(value) }
""")
        _julia_recursive_observation_expect_error(
            "source_location_nonprogress_direct_recursion",
            () -> runtime_parse(LinkedSpecRuntimeEngine(direct), "x"),
        )

        _, mutual = _julia_recursive_observation_compile("""Top:: I { value = observe_recognition(top_observation, call(MutualA)); return(value) }
MutualA: I { value = observe_recognition(observation_a, call(MutualB)); return(value) }
MutualB: I { value = observe_recognition(observation_b, call(MutualA)); return(value) }
""")
        _julia_recursive_observation_expect_error(
            "source_location_nonprogress_mutual_recursion",
            () -> runtime_parse(LinkedSpecRuntimeEngine(mutual), "x"),
        )
    end

    @testset "aborted observation propagates the original runtime diagnostic" begin
        _, aborted = _julia_recursive_observation_compile("""Top::
 I { value = observe_recognition(observation, call(AbortChild)); return(value) }
AbortChild:
 I { recognition_commit(missing) }
""")
        _julia_recursive_observation_expect_error(
            "recognition_token_expected",
            () -> runtime_parse(LinkedSpecRuntimeEngine(aborted), ""),
        )
    end

    @testset "generated plan executes the same private observation runtime" begin
        _, compiled = _julia_recursive_observation_compile(
            JULIA_RECURSIVE_OBSERVATION_SOURCE,
        )
        @test execute_generated_parser_v2(
            compiled,
            build_generated_rule_plan(compiled),
            "🙂",
            "recursive-observation/julia.spec",
        ) == Any[
            false,
            _julia_recursive_observation_expected(
                rule_label = "Child",
                invocation_id = 2,
                parent_invocation_id = 1,
                entry_offset = 0,
                selected_match = (0, 1),
                accepted_exit = 1,
                outcome = "accepted",
            ),
        ]
    end

    @testset "independently loaded emitted source uses the same private runtime" begin
        _, compiled = _julia_recursive_observation_compile(
            JULIA_RECURSIVE_OBSERVATION_SOURCE,
        )
        identity = "recursive-observation/julia-emitted.spec"
        value, source_identity = _julia_recursive_observation_execute_emitted(
            compiled,
            identity,
            "🙂",
        )
        @test value == Any[
            false,
            _julia_recursive_observation_expected(
                rule_label = "Child",
                invocation_id = 2,
                parent_invocation_id = 1,
                entry_offset = 0,
                selected_match = (0, 1),
                accepted_exit = 1,
                outcome = "accepted",
            ),
        ]
        @test source_identity == identity
    end
end

end
